### ISSUE TITLE
Wrong nfs host ip when using docker inside Vagrant

### ISSUE THREAD

**Issue by varac** [pradels/vagrant-libvirt#544]
Hi, 

i'm using NFS shared folders support for vagrant (libvirt provider). It works well, until i try to use docker inside a vagrant box.

Mounting the shared folder fails:  

```
==> jessie: Mounting NFS shared folders...
 INFO ssh: SSH is ready!
 INFO ssh: Execute:  (sudo=false)
 INFO guest: Execute capability: mount_nfs_folder [#<Vagrant::Machine: jessie (VagrantPlugins::ProviderLibvirt::Provider)>, "172.17.0.1", {"/srv/dev"=>{:nfs_version=>4, :mount_options=>["rw"], :guestpath=>"/srv/dev", :hostpath=>"/home/varac/dev", :disabled=>false, :__vagrantfile=>true, :map_uid=>1000, :map_gid=>1000, :nfs_udp=>true, :uuid=>"1541927476", :linux__nfs_options=>["rw", "no_subtree_check", "all_squash", "anonuid=1000", "anongid=1000", "fsid=1541927476"]}, "/vagrant"=>{:guestpath=>"/vagrant", :hostpath=>"/home/varac/thoughtworks/puppet-pixelated-vagrant/leap_platform", :disabled=>false, :__vagrantfile=>true, :map_uid=>1000, :map_gid=>1000, :nfs_udp=>true, :nfs_version=>3, :uuid=>"435293755", :linux__nfs_options=>["rw", "no_subtree_check", "all_squash", "anonuid=1000", "anongid=1000", "fsid=435293755"]}, "/tmp/vagrant-puppet/manifests-6aeec19aac243b610dabb6d3508d09c9"=>{:owner=>"root", :nfs__quiet=>true, :guestpath=>"/tmp/vagrant-puppet/manifests-6aeec19aac243b610dabb6d3508d09c9", :hostpath=>"/home/varac/thoughtworks/puppet-pixelated-vagrant/leap_platform/vagrant", :disabled=>false, :__vagrantfile=>true, :quiet=>true, :map_uid=>1000, :map_gid=>1000, :nfs_udp=>true, :nfs_version=>3, :uuid=>"3144777403", :linux__nfs_options=>["rw", "no_subtree_check", "all_squash", "anonuid=1000", "anongid=1000", "fsid=3144777403"]}, "/tmp/vagrant-puppet/modules-a9f1941cfdeaf1c1687a98098c6ad838"=>{:owner=>"root", :nfs__quiet=>true, :guestpath=>"/tmp/vagrant-puppet/modules-a9f1941cfdeaf1c1687a98098c6ad838", :hostpath=>"/home/varac/thoughtworks/puppet-pixelated-vagrant/leap_platform/puppet/modules", :disabled=>false, :__vagrantfile=>true, :quiet=>true, :map_uid=>1000, :map_gid=>1000, :nfs_udp=>true, :nfs_version=>3, :uuid=>"2817886613", :linux__nfs_options=>["rw", "no_subtree_check", "all_squash", "anonuid=1000", "anongid=1000", "fsid=2817886613"]}}] (debian)
 INFO ssh: SSH is ready!
 INFO ssh: Execute:  (sudo=false)
 INFO guest: Execute capability: shell_expand_guest_path [#<Vagrant::Machine: jessie (VagrantPlugins::ProviderLibvirt::Provider)>, "/srv/dev"] (debian)
 INFO ssh: Execute: echo; printf /srv/dev (sudo=false)
 INFO ssh: Execute: mkdir -p /srv/dev (sudo=true)
 INFO ssh: Execute: mount -o 'rw' 172.17.0.1:'/home/varac/dev' /srv/dev (sudo=true)

 INFO retryable: Retryable exception raised: #<Vagrant::Errors::LinuxNFSMountFailed: The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

mount -o 'rw' 172.17.0.1:'/home/varac/dev' /srv/dev

Stdout from the command:


Stderr from the command:

stdin: is not a tty
mount.nfs: Connection timed out
```

i'm using this vagrantfile:

```
--- puppet-pixelated-vagrant/leap_platform ‹develop* M› » cat Vagrantfile 
# -*- mode: ruby -*-
# vi: set ft=ruby :

Vagrant.configure("2") do |box|
  # Please verify the sha512 sum of the downloaded box before importing it into vagrant !
  # see https://leap.se/en/docs/platform/details/development#Verify.vagrantbox.download
  # for details


  box.vm.define :"jessie", primary: true do |config|

    config.vm.box = "LEAP/jessie"
    config.vm.provider "virtualbox" do |v|
      v.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
      v.name   = "jessie"
      v.memory = 1024
    end

    config.vm.provider "libvirt" do |v|
      v.memory = 1024
    end

    config.vm.provision "puppet" do |puppet|
      puppet.manifests_path = "./vagrant"
      puppet.module_path = "./puppet/modules"
      puppet.manifest_file = "install-platform.pp"
      puppet.options = "--verbose"
    end
    config.vm.provision "shell", path: "vagrant/configure-leap.sh"
    config.ssh.username = "vagrant"

    # forward leap_web ports
    config.vm.network "forwarded_port", guest: 8080,  host:8080
    config.vm.network "forwarded_port", guest: 443, host:4443
  end

  box.vm.define :"wheezy", autostart: false do |config|
    ...
  end
end
```

The problem here is, after docker is installed `vagrant reload` picks the network of the new `docker0` interface and tries to connect to the first IP in that net, but it should use `eth0` instead: 

```
vagrant@node1:~$ ps aux|grep mount
root      2361  0.0  0.0  21432   728 ?        S    10:32   0:00 mount -o rw 172.17.0.1:/home/varac/dev /srv/dev
root      2362  0.0  0.1  40152  1316 ?        S    10:32   0:00 /sbin/mount.nfs 172.17.0.1:/home/varac/dev /srv/dev -o rw


vagrant@node1:~$ ip r
default via 192.168.121.1 dev eth0 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.42.1 
192.168.121.0/24 dev eth0  proto kernel  scope link  src 192.168.121.150 


vagrant@node1:~$ /sbin/ifconfig 
docker0   Link encap:Ethernet  HWaddr 56:84:7a:fe:97:99  
          inet addr:172.17.42.1  Bcast:0.0.0.0  Mask:255.255.0.0
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eth0      Link encap:Ethernet  HWaddr 52:54:00:5f:06:f2  
          inet addr:192.168.121.150  Bcast:192.168.121.255  Mask:255.255.255.0
          inet6 addr: fe80::5054:ff:fe5f:6f2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:9847 errors:0 dropped:2 overruns:0 frame:0
          TX packets:3928 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:5875422 (5.6 MiB)  TX bytes:330494 (322.7 KiB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:26347 errors:0 dropped:0 overruns:0 frame:0
          TX packets:26347 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:2108985 (2.0 MiB)  TX bytes:2108985 (2.0 MiB)

```

i tried to debug the code in `lib/vagrant-libvirt/action/prepare_nfs_settings.rb`, 
and indeed, it's evaluated falsely in there. I changed the line 25 from

```
@logger.info("host IP: #{env[:nfs_host_ip]} machine IP: #{env[:nfs_machine_ip]}")
```
to 

```
puts "host IP: #{env[:nfs_host_ip]} machine IP: #{env[:nfs_machine_ip]}"
```

and indeed, during a `vagrant reload`:
```
...
host IP: 172.17.0.1 machine IP: ["192.168.121.150", "172.17.42.1"]
...
```

it should pick the latter one.

**Comment by varac** [comment]
btw, this is reproducible by the cmd documented here: https://github.com/pixelated/puppet-pixelated#testing-pixelated

**Comment by varac** [comment]
until this is fixed, let's apply a dirty hotfix here:

```
    diff prepare_nfs_settings.rb.orig prepare_nfs_settings.rb
    45c45
    <             s.connect(ip.last, 1)
    ---
    >             s.connect(ip.first, 1)
    >

```

**Comment by infernix** [comment]
I had a look at this and there was a change in functionality in PR #463 but there's no clear reason as to why.

I tried to reproduce your leap platform build but that failed (see below).

That being said, this seems to be a very crude way of determining the host ip by just listing all possible IPs on a host and picking whichever's last. I would imagine that  `def read_host_ip` could simply get the host IP through libvirts network definition by using `virNetworkGetXMLDesc` for the active network instead of assuming that the last IP in the array is the correct host network.

And if I'm not mistaken this is already available in `libvirt_networks` in `network_util.rb`- we would just need to pass in the machines (list of) networks.

```
Bringing machine 'jessie' up with 'libvirt' provider...
...
==> jessie:  = Updating submodule puppet/modules/apache
==> jessie:  = run git submodule update --init puppet/modules/apache 2>&1
==> jessie:    - FAILED (exit 1) fatal: Not a git repository: /tmp/9p/leap_platform/.git/modules/puppet/modules/apache
==> jessie: Unable to find current revision in submodule path 'puppet/modules/apache'
==> jessie:  = bailing out 
==> jessie: [master e78a8be] initialized and deployed provider
==> jessie:  Committer: vagrant <vagrant@leap-jessie.local>
==> jessie: Your name and email address were configured automatically based
==> jessie: on your username and hostname. Please check that they are accurate.
==> jessie: You can suppress this message by setting them explicitly:
==> jessie: 
==> jessie:     git config --global user.name "Your Name"
==> jessie:     git config --global user.email you@example.com
==> jessie: 
==> jessie: After doing this, you may fix the identity used for this commit with:
==> jessie: 
==> jessie:     git commit --amend --reset-author
==> jessie: 
==> jessie:  1 file changed, 1 insertion(+)
==> jessie:  create mode 100644 facts.json
==> jessie: Failed to stop leap-mx.service: Unit leap-mx.service not loaded.
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

**Comment by jim-minter** [comment]
Hmm, I've had a bit more of a look at this and prepare_nfs_settings.rb is a bit grotty!

1. a reader might think `def ping()` would use ICMP to check VM accessibility, but in fact it attempts to connect via TCP to the echo service (port 7).  It accepts ECONNREFUSED as indicating the VM is up, but many VM Linux firewalls will firewall port 7 by default and cause EHOSTUNREACH to be returned on the host.  I suspect this makes this "ping" method unable to differentiate between a firewalled VM being otherwise accessible, and a VM being inaccessible.  If nothing else, checking on port 22 would surely be better than 7.

2. If a VM doesn't firewall port 7, `def read_machine_ip()` will return `machine.ssh_info`, which IMO is good.

3. Otherwise, `def read_machine_ip()` goes through a bit of a mad process to enumerate VM IPs and try to "ping" them from the host.

4. If no IPs "ping" successfully, `def read_machine_ip()` returns **all** of them rather than an empty list.  IMO this is pretty bad.

5. As a result, `def read_host_ip()` is geared optionally to accept arrays as input, which I think is a grim hack.

6. A fair amount of this code dates from 0deb639 / #366 , which to me seems a very specific use case which (a) I think makes it hard to refactor this code while guaranteeing not to break it, and (b) I think might be better handled not as a vagrant-libvirt feature but using a custom provision script anyway.

TL;DR: I would aim to refactor this script by removing `def ping()`, setting something like:

```ruby
        def read_machine_ip(machine)
          machine.ssh_info[:host]
```

```ruby
        def read_host_ip(ip)
          UDPSocket.open do |s|
            s.connect(ip, 1)
            s.addr.last
          end
        end
```

However, I'd be interested in others' views!

**Comment by miminar** [comment]
Same problem here.

**Comment by bspink** [comment]
Hitting this problem as well.

**Comment by jwmatthews** [comment]
I hit same problem with vagrant-libvirt 0.37 with docker present in our VMs when using NFS.

Following @varac's comment and @jim-minter's notes I changed the below 2 lines and had success.

$ diff prepare_nfs_settings.rb prepare_nfs_settings.rb.ORIGINAL 
25d24
< 
46c45
<               s.connect(ip.first, 1)
---
>               s.connect(ip.last, 1)
61c60
<           return ssh_host
---
>           return ssh_host if ping(ssh_host)

**Comment by aarondl** [comment]
Is there any better way to set these manually than by editing source code?

**Comment by narkisr** [comment]
Same issue picking the last one arbitrary does not make sense

**Comment by witjoh** [comment]
Just some info

Dont think ip.last or ip.first is a good solution.  I'm running docker and vagrant-libvirt inside a libvirt box, and that creates : 
docker0 : 172....
eth0 : 192.168.121 .....
virbr0: 192.168.122 .....
and in version 2.0.0 the export is define on the virbr0.   For now I just edit the /etc/exports on the host.
I cannot find any config setting to force eth0 as ip tu use in the /etc/exports
